madanalysis/configuration/weight_configuration.py

```diff
@@ -0,0 +1,303 @@
+
+        self._names = [x.name for x in self]
+        return self._names
+
+    def to_dict(self) -> List[Dict[Text, Any]]:
+        """Convert to dictionary"""
+        return [x.to_dict() for x in self]
+
+    def from_dict(self, data: List[Dict[Text, Any]]) -> None:
+        """Construct collection from data"""
+        for dat in data:
+            self.append(dat["name"], dat["loc"])
+
+    @property
+    def nominal(self) -> Weight:
+        """Get nominal weight"""
+        for w in self:
+            if w.is_nominal:
+                return w
+        raise ValueError("Can not find nominal weight")
+
+    def group_for(self, group: Text) -> Dict:
+        """Create a group"""
+        assert group in ["muf", "mur", "pdfset", "dynamic_scale"]
+
+        unique = np.unique(
+            [getattr(w, group) for w in self if getattr(w, group) is not None]
+        )
+
+        if len(unique) == 0:
+            return {}
+
+        group_dict = {v: [] for v in unique}
+        for w in self:
+            if getattr(w, group) is not None:
+                group_dict[getattr(w, group)].append(w)
+
+        return group_dict
+
+    def pdfset(self, pdfid: int) -> List[Weight]:
+        """retreive weights corresponding to one pdf set"""
+        return WeightCollection([w for w in self if w.pdfset == pdfid])
+
+    def get(self, **kwargs) -> List[Weight]:
+        if len(kwargs) == 0:
+            return []
+        collection = []
+        keys = ["muf", "mur", "dynamic_scale", "pdfset", "merging"]
+
+        for weight in self:
+            add = True
+            for key in (key for key in keys if key in kwargs):
+                if getattr(weight, key) != kwargs.get(key):
+                    add = False
+                    break
+            if add:
+                collection.append(weight)
+
+        return WeightCollection(collection)
+
+    @property
+    def pdfsets(self) -> List[int]:
+        """Retreive a list of pdfsets"""
+        return np.unique([w.pdfset for w in self if w.pdfset is not None]).tolist()
+
+    @property
+    def scales(self) -> Dict[Text, List[float]]:
+        """return scale for muf and mur"""
+        muf = np.unique([w.muf for w in self if w.muf is not None]).tolist()
+        muf.sort()
+        mur = np.unique([w.mur for w in self if w.mur is not None]).tolist()
+        mur.sort()
+        return {"muf": muf, "mur": mur}
+
+    @property
+    def has_scale(self) -> bool:
+        """is there any scale variations"""
+        return len(self.scales["muf"]) > 0
+
+    @property
+    def central_scale(self) -> float:
+        """retreive central scale"""
+        scales = self.scales["muf"]
+        return scales[len(scales) // 2]
+
+    def get_scale_vars(self, point: int = 3, dynamic: int = None) -> Tuple:
+        if dynamic is not None:
+            dynamic = dynamic if self.has_dyn_scale(dynamic) else None
+
+        scales = self.scales["muf"]
+        if len(scales) == 5 and point == 3:
```

Comment:
I am not too sure to understand how we could get 5 entries for the scale variations. Also, what does the code do if more entries are available (as the user may want to have a custom way to evaluate scale variations). I would like to see a default: if we have neither 3, 7 or 9 point scale variations, then we just take all that is available.

Reply:
this is the custom user will choose 3,7 or 9 and depending on how many there are within the file it will return the result accordingly. I forgot 5 was not the case

Reply:
I have fixed this part of the code locally. Commit is coming.